### DISCUSSION TITLE
Creating mock 'stubs'

### DISCUSSION THREAD

**Discussion by bugbountyguy** [MikeSchulze/gdUnit4#321] [Q&A] [answered]
Hi,

I'm trying to write some unit tests that should basically 'intercept' some file system operations. I have a function that has some logic in it that needs testing, which includes calling a locally instantiated ConfigFile's save() if a certain condition is false. However, I can't seem to find a way to mock that save function to _not_ save an actual file to the system. 

So basically I'm trying to:
1. Stub out the save() to just return OK regardless of what's passed in
2. Run my test with the condition that would result in it being false
3. Assert that the save function was called with a string

I tried to look at the 'mocks' documentation, but couldn't figure out a solution (especially since the config file is instantiated within the function's scope). 

Any suggestions on how I can implement this other than making the config file variable scoped 'publicly' to replace with a mock object in my class?

edit:
I reorganized the code a little bit, and tried doing a mock. It appears the mocks don't work though, so maybe this is a bug I need to report instead.
```
var mock := mock(MySingletonClass) as MySingletonClass
do_return(OK).on(mock)._save_config(any)
...
```
I get an error saying `'Invalid call. Nonexistent function '_save_config' in base 'Nil'.` - I'm wondering if I'm running into issues here because it's a singleton with autoload. So I guess I'm now having 2 different issues lol.

**Comment by MikeSchulze** [comment, marked as answer]
Hi @bugbountyguy 

I assume you are using Godot `ConfigFile`, the core function `save`.
With Godot version 4.x, the GDScript has been rewritten so that overwriting a core function is no longer possible, which breaks my mock implementation.
For more details, see https://github.com/MikeSchulze/gdUnit4/issues/42

According to the Godot core developers, overwriting core functions was a bug. 
To solve this problem, I have some ideas, but the implementation will take some time.
I have checked the GdUnit4 documentation, and I am missing this important information about mocking core functions, I
will add it.

**Reply by bugbountyguy** [reply]
ahhh man that's too bad. Thank you for confirming - I was going insane thinking I did something wrong, lol. I'll follow their issue - thank you!
